File: c++/062.cpp
class Solution {
/*
1. 若把向下看做“写0”，把向上看做“写1”，则问题即为：求m-1个0，n-1个1的全排列个数
2. 公式为：(m+n-2)! / ( (m-1)! * (n-1)! )
3. 难点在于：直接算阶乘必越界，故需要利用“/”来减小每一步的 r 和 i

公式中（m+n-2）> max(m-1,n-1)，故阶乘可抵消max

进阶：可在抵消掉最大值之后，从小到大同时乘除



动态规划：dp[i][j] = dp[i-1][j] + dp[i][j-1]
即：每一步的可从上一格或左一格走来，故每一格的路径数即为上一格的路径数 + 左一格的路径数
而初始化：即边界点均为1

*/
public:
    int uniquePaths(int m, int n) {
        if(m==1||n==1) return 1;
        int max,min;
        max = m>n?m:n;
        min = m>n?n:m;
        int r=1, Min[min];
        for(int i=2; i<min; i++)
            Min[i]=i;
        for(int i=m+n-2;i>=max; i--){
            int t=i;
            for(int j=2; j<min; j++){
                if(Min[j]>1&&t%Min[j]==0){
                    t /= Min[j];
                    Min[j] = 1;
                }
                if(Min[j]>1&&r%Min[j]==0){
                    r /= Min[j];
                    Min[j] = 1;
                }
            }
            r *= t;
        }
        for(int j=2; j<min; j++){
            if(Min[j]>1&&r%Min[j]==0){
                r /= Min[j];
                Min[j] = 1;
            }
        }
        return r;
    }
};

File: go/062.go
package main

/**
不同路径
in:  m = 3, n = 7
out: 28
思路：到达最后一步的路径为上面和左边之和
动态规划 f[m][n] 状态：00 - mn的路径  属性：数量
转移方程：f[m][n] = f[m-1][n] + f[m][n-1]
初值：f[0][i] f[i][0] = 1
*/
func uniquePaths(m int, n int) int {
	f := make([][]int, m)
	for i := 0; i < len(f); i++ {
		f[i] = make([]int, n)
	}
	for i := 0; i < m; i++ {
		f[i][0] = 1
	}
	for i := 0; i < n; i++ {
		f[0][i] = 1
	}
	for i := 1; i < m; i++ {
		for j := 1; j < n; j++ {
			f[i][j] = f[i-1][j] + f[i][j-1]
		}
	}
	return f[m-1][n-1]
}
